backend/core/src/main/java/org/sonarsource/sonarlint/core/fs/FileExclusionService.java

```diff
@@ -185,7 +192,120 @@ public boolean isExcluded(URI fileUri) {
     return Boolean.TRUE.equals(serverExclusionByUriCache.get(fileUri));
   }
 
-  public List<URI> filterOutClientExcludedFiles(String configurationScopeId, List<URI> files) {
+  public List<ClientFile> refineAnalysisScope(AnalysisTask task, Path baseDir) {
+    var configScopeId = task.getConfigScopeId();
+    var requestedFileUris = task.getFilesToAnalyze();
+    if (task.getTriggerType().equals(TriggerType.FORCED)) {
+      var filteredURIsNoFile = new ArrayList<URI>();
+      var filesToAnalyze = requestedFileUris.stream().map(uri -> {
+        var file = findFile(configScopeId, uri);
+        if (file == null) {
+          filteredURIsNoFile.add(uri);
+        }
+        return file;
+      })
+        .filter(Objects::nonNull)
+        .toList();
+      logFilteredURIs("Filtered out URIs having no file", filteredURIsNoFile);
+      return filesToAnalyze;
+    }
+    return filterOutExcludedFiles(configScopeId, baseDir, requestedFileUris);
+  }
+
+  private List<ClientFile> filterOutExcludedFiles(String configurationScopeId, Path baseDir, List<URI> files) {
+    var sonarLintGitIgnore = createSonarLintGitIgnore(baseDir);
+    // INFO: When there are additional filters coming at some point, add them here and log them down below as well!
+    var filteredURIsFromExclusionService = new ArrayList<URI>();
+    var filteredURIsFromGitIgnore = new ArrayList<URI>();
+    var filteredURIsNotUserDefined = new ArrayList<URI>();
+    var filteredURIsFromSymbolicLink = new ArrayList<URI>();
+    var filteredURIsFromWindowsShortcut = new ArrayList<URI>();
+    var filteredURIsNoFile = new ArrayList<URI>();
+
+    // Do the actual filtering and in case of a filtered out URI, save them for later logging!
+    var actualFilesToAnalyze = filterOutClientExcludedFiles(configurationScopeId, files)
+      .stream()
+      .map(uri -> {
+        var file = findFile(configurationScopeId, uri);
+        if (file == null) {
+          filteredURIsNoFile.add(uri);
```

Comment:
Important part of functionality is not tested, to be tested later..

Reply:
Ah yes, indeed there is a lack of coverage here. I will cover it later